src/quadiron_c.cpp

```diff
@@ -0,0 +1,382 @@
+    std::vector<bool> wanted_idxs_vec(fec->n_outputs, false);
+    int metadata_size = quadiron_fnt32_get_metadata_size(fecp, block_size);
+    bool res;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[i].fnt_deserialize(
+                    metadata, metadata_size / 4);
+                if (ret == -1)
+                    return -1;
+            }
+            parities_vec[i] = parity[i] + metadata_size;
+        }
+    } else {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            if (!missing_idxs[i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(data[i]);
+                int ret = parities_props[i].fnt_deserialize(
+                    metadata, metadata_size / 4);
+                if (ret == -1)
+                    return -1;
+            }
+            parities_vec[i] = data[i] + metadata_size;
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[fec->n_data + i].fnt_deserialize(
+                    metadata, metadata_size / 4);
+                if (ret == -1)
+                    return -1;
+            }
+            parities_vec[fec->n_data + i] = parity[i] + metadata_size;
+        }
+    }
+
+    int need_decode = 0;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        /*
+         * Easy case where the target is a data then simply decode
+         */
+        if (destination_idx < fec->n_data) {
+            wanted_idxs_vec[destination_idx] = true;
+            res = fec->decode_blocks_vertical(
+                data_vec,
+                parities_vec,
+                parities_props,
+                missing_idxs_vec,
+                wanted_idxs_vec,
+                block_size);
+            if (!res) {
+                return -1;
+            }
+            return 0;
+        }
+    }
+
+    /*
+     * At this point we want a parity to be reconstructed
+     * If systematic we may need to decode if a data is missing.
+     * If non-systematic we always need to decode
+     */
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            if (missing_idxs[i]) {
+                need_decode = 1;
+                wanted_data_vec[i] = true;
+                std::vector<uint8_t>* block =
+                    new std::vector<uint8_t>(block_size);
+                data_vec[i] = block->data();
```

Comment:
@slaperche-scality when I allocate in the stack it does not work.

std::vector<uint8_t> block(block_size);

The liberasurecode_test does not pass. I don't understand since those buffers are not meant to be returned. Is it because block_size can be very large ?